### PR TITLE
Ready for Review: Check for valid Name in UI

### PR DESCRIPTION
**Pull request by bliemelh**
https://trello.com/c/j0JTyQ1S/314-invalide-zeichen-im-modellnamen
not done yet:
check import for invalid name,
show message if name is invalid

### CODE REVIEW

**Comment by junkerm** [comment]
It would be good to have an error message similar to the one in the editors, when the name is not valid. 
image

**Comment by junkerm** [comment]
I would suggest a different error message, e.g. saying that which characters are invalid
image

**Comment by junkerm** [comment]
When opening a CEG Editor and adding and selecting a node, the field condition is missing and an error message is shown:

image

image

File: web/src/app/modules/forms/modules/generic-form/components/generic-form.component.ts [hunk 1]
```diff
@@ -1,166 +1,172 @@
...
+        this.orderFieldMeta();
+        let formBuilderObject: any = {};
+        for (let i = 0; i < this._meta.length; i++) {
+            let fieldMeta: FieldMetaItem = this._meta[i];
+            let fieldName: string = fieldMeta.name;
+            let formBuilderObjectValue: any[] = [''];
+            const validators = [];
+            if (this._meta[i].required) {
+                validators.push(Validators.required);
+                // formBuilderObjectValue.push(Validators.required);
```
Comment (junkerm) [inline web/src/app/modules/forms/modules/generic-form/components/generic-form.component.ts]: please remove commented code

File: web/src/app/modules/forms/modules/generic-form/components/generic-form.component.ts [hunk 2]
```diff
@@ -1,166 +1,172 @@
...
+            }
+            if (this._meta[i].allowedPattern) {
+                validators.push(Validators.pattern(new RegExp('^((?![,;\|]).)*$')));
```
Comment (junkerm) [inline web/src/app/modules/forms/modules/generic-form/components/generic-form.component.ts]: better use the pattern from the meta information

File: web/src/app/modules/views/main/editors/modules/contents-container/base/contents-container-base.ts [hunk 3]
```diff
@@ -1,117 +1,119 @@
...
+        protected modal: ConfirmationModal,
+        private clipboardService: ClipboardService) {
+    }
+
+    public async ngOnInit(): Promise<void> {
+        await this.readContents();
+    }
+
+    public async create(name: string): Promise<void> {
+        let invalidCharacters: RegExp = new RegExp('[,;|]');
```
Comment (junkerm) [inline web/src/app/modules/views/main/editors/modules/contents-container/base/contents-container-base.ts]: Can you reuse the pattern configured in the field meta information, e.g. by FieldMeta.INamed.allowedPattern?

File: web/src/app/modules/views/main/editors/modules/contents-container/base/contents-container-base.ts [hunk 4]
```diff
@@ -1,117 +1,119 @@
...
+    public async copy(model: IContainer): Promise<void> {
+        const copy = Objects.clone(model) as IContainer;
+        copy.description = model.description;
+        copy.id = Id.uuid;
+        copy.name = this.translate.instant('copyOf') + ' ' + model.name;
+        this.clipboardService.clipboard = [copy];
+    }
+
+    public async paste(name: string): Promise<void> {
+        let invalidCharacters: RegExp = new RegExp('[,;|]');
```
Comment (junkerm) [inline web/src/app/modules/views/main/editors/modules/contents-container/base/contents-container-base.ts]: same as above, use the defined pattern from the ecore model

File: web/src/app/modules/views/main/editors/modules/contents-container/components/ceg-model-container.component.html [hunk 5]
```diff
@@ -1,69 +1,70 @@
...
+                        </button>
+                    </div>
+                </div>
+            </div>
+            <div class="col">
+                <div class="input-group">
+                    <input type="text" #pasteNameInput (keyup)="0" class="form-control form-control-sm" [placeholder]="clipboardModelName" [disabled]="!canPaste" />
+                    <div class="input-group-append">
+                        <button id="requirement-paste-button" title="{{'paste' | translate}}" (click)="paste(pasteNameInput.value)" class="btn btn-sm btn-outline-primary"
+                            [disabled]="!canPaste || pasteNameInput.value.match('[,;|]')"
```
Comment (junkerm) [inline web/src/app/modules/views/main/editors/modules/contents-container/components/ceg-model-container.component.html]: the regexp check should not be defined here in the template

File: web/src/app/modules/views/main/editors/modules/contents-container/components/ceg-model-container.component.ts [hunk 6]
```diff
@@ -1,42 +1,50 @@
...
+    protected condition = (element: IContainer) => Type.is(element, CEGModel);
+
+    public async createElement(name: string): Promise<CEGModel> {
+        let factory: ModelFactoryBase = new CEGModelFactory(this.dataService);
+        const element = await factory.create(this.parent, true, Id.uuid, name);
+        return element as CEGModel;
+    }
+
+    public isCreateCEGButtonEnabled(name: string): boolean {
+        let invalidCharacters: RegExp = new RegExp('[,;|]');
```
Comment (junkerm) [inline web/src/app/modules/views/main/editors/modules/contents-container/components/ceg-model-container.component.ts]: same as above... as this sees to be stereotypic code, you could think about extracting it to a utility class

File: web/src/app/modules/views/main/editors/modules/contents-container/components/folder-container.component.ts [hunk 7]
```diff
@@ -1,43 +1,51 @@
...
+        return element;
+    }
+
+    public async delete(element: Folder): Promise<void> {
+        const message = this.translate.instant('doYouReallyWantToDeleteFolder', { name: element.name });
+        await super.delete(element, message);
+    }
+
+    public isCreateFolderButtonEnabled(name: string): boolean {
+        let invalidCharacters: RegExp = new RegExp('[,;|]');
```
Comment (junkerm) [inline web/src/app/modules/views/main/editors/modules/contents-container/components/folder-container.component.ts]: see comments above

File: web/src/app/modules/views/main/editors/modules/contents-container/components/process-model-container.component.html [hunk 8]
```diff
@@ -1,70 +1,71 @@
...
+                </div>
+            </div>
+            <div class="col">
+                <div class="input-group">
+                    <input type="text" #pasteNameInput (keyup)="0" class="form-control form-control-sm" [placeholder]="clipboardModelName"
+                        [disabled]="!canPaste" />
+                    <div class="input-group-append">
+                        <button id="requirement-paste-button" title="{{'paste' | translate}}" (click)="paste(pasteNameInput.value)"
+                            class="btn btn-sm btn-outline-primary" [disabled]="!canPaste || pasteNameInput.value.match('[,;|]')"
+                            [class.disabled]="!canPaste || pasteNameInput.value.match('[,;|]')">
```
Comment (junkerm) [inline web/src/app/modules/views/main/editors/modules/contents-container/components/process-model-container.component.html]: don't do matching in the template

File: web/src/app/modules/views/main/editors/modules/contents-container/components/process-model-container.component.ts [hunk 9]
```diff
@@ -1,42 +1,50 @@
...
+    protected condition = (element: IContainer) => Type.is(element, Process);
+
+    public async createElement(name: string): Promise<Process> {
+        let factory: ModelFactoryBase = new ProcessFactory(this.dataService);
+        const element = await factory.create(this.parent, true, Id.uuid, name);
+        return element as Process;
+    }
+
+    public isCreateProcessButtonEnabled(name: string): boolean {
+        let invalidCharacters: RegExp = new RegExp('[,;|]');
```
Comment (junkerm) [inline web/src/app/modules/views/main/editors/modules/contents-container/components/process-model-container.component.ts]: see comments above

File: web/src/app/modules/views/main/editors/modules/contents-container/components/test-specification-container.component.ts [hunk 10]
```diff
@@ -1,54 +1,62 @@
...
+    }
+
+    protected async readContents(): Promise<void> {
+        const contents = await this.dataService
+            .performQuery(this.parent.url, 'listRecursive', { class: TestSpecification.className });
+        this.contents = Sort.sortArray(contents);
+    }
+
+    public isCreateTestspecificationButtonEnabled(name: string): boolean {
+        let invalidCharacters: RegExp = new RegExp('[,;|]');
```
Comment (junkerm) [inline web/src/app/modules/views/main/editors/modules/contents-container/components/test-specification-container.component.ts]: see comments above

File: web/src/app/validation/valid-name-validator.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+import { TranslateService } from '@ngx-translate/core';
+
+export class ValidNameValidator extends ElementValidatorBase<IContainer> {
+
+    constructor() {
+        super();
+    }
+
+    public validate(element: IContainer, contents: IContainer[] = []): ValidationResult {
+       let invalidCharacters: RegExp = new RegExp('[,;|]');
```
Comment (junkerm) [inline web/src/app/validation/valid-name-validator.ts]: see comments above

File: bundles/specmate-model-ecore/model/specmate.ecore [hunk 12]
```diff
@@ -10,6 +10,7 @@
           <details key="required" value="true"/>
           <details key="type" value="text"/>
           <details key="position" value="0"/>
+          <details key="allowedPattern" value="^((?![,;\|]).)*$"/>
```
[the remark this answers is not in the document]
Comment (bliemelh) [reply bundles/specmate-model-ecore/model/specmate.ecore]: yes, that's better.

File: web/src/app/modules/views/main/editors/modules/contents-container/base/contents-container-base.ts (hunk 3 above)
Comment (bliemelh) [reply web/src/app/modules/views/main/editors/modules/contents-container/base/contents-container-base.ts]:
```suggestion
        let validPattern = MetaInfo.INamed[0].allowedPattern;
```
is this alright?

File: web/src/app/validation/valid-name-validator.ts (hunk 11 above)
Comment (bliemelh) [reply web/src/app/validation/valid-name-validator.ts]:
```suggestion
       let validPattern = MetaInfo.INamed[0].allowedPattern;
       let validName: RegExp = new RegExp(validPattern);
```

File: web/src/app/modules/forms/modules/generic-form/base/form-element.ts [hunk 13]
```diff
@@ -1,19 +1,29 @@
...
+    @Input()
+    public meta: FieldMetaItem;
+
+    @Input()
+    public form: FormGroup;
+
+    constructor(private translate: TranslateService) { }
+
+    public get errorMessage(): string {
+        if (this.form.controls.name != undefined && this.form.controls.name.errors != undefined) {
```
Comment (junkerm) [inline web/src/app/modules/forms/modules/generic-form/base/form-element.ts]:
```suggestion
         // fetch the control that actually represents this field
         const control = this.form.controls[this.meta.name]; 
         
        if (control != undefined && control.errors != undefined) {
```

Comment (junkerm) [reply web/src/app/modules/forms/modules/generic-form/base/form-element.ts]: I discussed this with @sebeder  and we think this should be the valid solution

File: web/src/app/modules/forms/modules/generic-form/base/form-element.ts [hunk 14]
```diff
@@ -1,19 +1,29 @@
...
+            let error = this.form.controls.name.errors;
```
Comment (junkerm) [inline web/src/app/modules/forms/modules/generic-form/base/form-element.ts]:
```suggestion
            let error = control.errors;
```

**Pull request merged**